File: Lista01/batalhaNaval/Jogador.java
package lista1.batalhaNaval;

public class Jogador{
    char[][] mapa = new char[5][5];
    String nome;
    final int MAX_COORD = 4;
    final int MIN_COORD = 0;
    final char VAZIO = '-';
    final char NAVIO = '@';
    final char AFUNDADO = 'X';
    final char BOMBA = 'O';
    boolean ganhou = false;
    short qtdNaviosPerdidos = 0;
    public Jogador(int id){
        // inicializa o mapa
        this.inicializaMapa();
        
        // nome de player
        this.nome = "Player " + id;
    }

    public void inicializaMapa(){
        for (short x = 0; x <= this.MAX_COORD; x++)
            for (short y = 0; y<=this.MAX_COORD; y++) this.mapa[x][y] = this.VAZIO;
    }

    public String insereCoordenadas(int tempx, int tempy){
        boolean existe;
        // validar o máximo
        if (tempx > this.MAX_COORD || tempy > this.MAX_COORD || tempx < this.MIN_COORD || tempy < this.MIN_COORD){
            return "Invalid coordinates. Choose different coordinates.";
        } else {
            existe = false;
            // valida com o mapa
            if (this.mapa[tempx][tempy] == '@'){
                existe = true;
            }
            
            if (existe) {
                return "You already have a ship there. Choose different coordinates.";
            } else {
                this.mapa[tempx][tempy]=this.NAVIO;
                return "";
            }
        }
    }

    public void exibeMapaPlayer(){
        System.out.println("\n\n  0  1  2  3  4");
        for (short x = 0; x <= this.MAX_COORD; x++){
            System.out.print(x);
            for (short y = 0; y<=this.MAX_COORD; y++){
                System.out.print(" "+this.mapa[x][y]+" ");
            }
            System.out.println();
        }
    }

    public void exibeMapaGame(){
        char temp;
        System.out.println("\n\n  0  1  2  3  4");
        for (short x = 0; x <= this.MAX_COORD; x++){
            System.out.print(x);
            for (short y = 0; y<=this.MAX_COORD; y++){
                temp = this.mapa[x][y];
                if (temp == this.NAVIO) System.out.print(" "+ this.VAZIO +" ");
                else System.out.print(" "+this.mapa[x][y]+" ");
            }
            System.out.println();
        }
        System.out.println();
    }

    public boolean atiraEm(Jogador alvo, short tempx, short tempy){
        // validar o máximo
        if (tempx > this.MAX_COORD || tempy > this.MAX_COORD || tempx < this.MIN_COORD || tempy < this.MIN_COORD){
            System.out.println("Invalid coordinates. Choose different coordinates.");
            return false;
        }

        char ponto = alvo.mapa[tempx][tempy];

        if (ponto == this.NAVIO){
            alvo.mapa[tempx][tempy] = this.AFUNDADO;
            System.out.println(this.nome + " hit " + alvo.nome + "'s Ship!!!");
            alvo.qtdNaviosPerdidos ++;
            if (alvo.qtdNaviosPerdidos == App4.QTD_NAVIOS) this.ganhou = true;
            return true;
        } else if(ponto == this.VAZIO) {
            alvo.mapa[tempx][tempy] = this.BOMBA;
            System.out.println(this.nome + ", MISSED!");
            return true;
        }  else if (ponto == this.BOMBA || ponto == this.AFUNDADO) {
            System.out.println("You already fired on this spot. Choose different coordinates.");
            return false;
        }
        return true;
    }
 }

File: Lista02/batalhaNaval/jogador.py
class Jogador:
    qtdNavios = 5

    def __init__(self, id):
        self.__mapa = list()
        self.__nome = f'Player {id}'
        self.__MAX_COORD = 4
        self.__MIN_COORD = 0
        self.__VAZIO = '-'
        self.__NAVIO = '@'
        self.__AFUNDADO = 'X'
        self.__BOMBA = 'O'
        self.__ganhou = False
        self.__qtdNaviosPerdidos = 0

        self.inicializaMapa()

    def getNome(self):
        return self.__nome
    
    def ganhou(self):
        return self.__ganhou

    def getNaviosPerdidos(self):
        return self.__qtdNaviosPerdidos
    
    def getPonto(self, x, y):
        return self.__mapa[x][y]

    def inicializaMapa(self):
        # cria o mapa vazio
        for i in range(self.__MAX_COORD+1):
            self.__mapa.append([self.__VAZIO]*(self.__MAX_COORD+1))

    def insereCoordenadas(self, tempx, tempy):
        # valida o máximo e mínimo
        if tempx > self.__MAX_COORD or tempy > self.__MAX_COORD or tempx < self.__MIN_COORD or tempy < self.__MIN_COORD:
            return 'Invalid coordinates. Choose different coordinates.'
        else:
            existe = False
            # valida com o mapa
            if self.__mapa[tempx][tempy] == self.__NAVIO:
                existe = True

            if existe:
                return 'You already have a ship there. Choose different coordinates.'
            else:
                self.__mapa[tempx][tempy]=self.__NAVIO
                return ''
            
    def exibeMapaPlayer(self):
        # exibe como está no mapa
        print('\n\n  0  1  2  3  4')
        for i in range(self.__MAX_COORD+1):
            print(i, end='')
            for j in range(self.__MAX_COORD+1):
                print(f" {self.__mapa[i][j]} ", end='')
            print()

    def exibeMapaGame(self):
        # esconde os navios
        print('\n\n  0  1  2  3  4')
        for i in range(self.__MAX_COORD+1):
            print(i, end='')
            for j in range(self.__MAX_COORD+1):
                temp = self.__mapa[i][j]
                if temp == self.__NAVIO:
                    print(f' {self.__VAZIO} ', end='')
                else:
                    print(f' {self.__mapa[i][j]} ', end='')
            print()
    
    def perdeuNavio(self, x, y):
        # troca no mapa
        self.__mapa[x][y] = self.__AFUNDADO
        # atualiza a variável
        self.__qtdNaviosPerdidos+=1

    def tiroBomba(self, x, y):
        # troca no mapa
        self.__mapa[x][y] = self.__BOMBA
        

    def atiraEm(self, alvo, tempx, tempy):
        # validar o máximo
        if tempx > self.__MAX_COORD or tempy > self.__MAX_COORD or tempx < self.__MIN_COORD or tempy < self.__MIN_COORD:
            print('Invalid coordinates. Choose different coordinates.')
            return False
        
        # pega o ponto no mapa
        ponto = alvo.getPonto(tempx, tempy)

        # se for navio, navio afunda
        if ponto == self.__NAVIO:
            alvo.perdeuNavio(tempx, tempy)
            print(f'{self.__nome} hit {alvo.getNome()}\'s Ship!!!')
            if alvo.getNaviosPerdidos() == Jogador.qtdNavios:
                self.__ganhou = True
            return True
        
        # se for espaço vazio, so estoura a bomba
        elif ponto == self.__VAZIO:
            alvo.tiroBomba(tempx, tempy)
            print(f'{self.__nome}, MISSED!')
            return True
        
        # se for um ponto já atingido, outro ponto será pedido do jogador
        elif ponto == self.__BOMBA or ponto == self.__AFUNDADO:
            print('You already fired on this spot. Choose different coordinates.')
            return False
